### PR TITLE
Streamlining CI/CD workflow for iOS Build and Release Process in Talawa

### LINKED ISSUE #2136: Generate an .IPA file and link it to README.md
**Is your feature request related to a problem? Please describe.**

There currently exists a CI/CD  pipeline for generating Talawa .apk [artifacts](https://github.com/PalisadoesFoundation/talawa/commit/9dac05567a1afdb8c1191ecd1d6f5b6bdc4af6f0#diff-1790f2f8b7123d50d46235053b54f1cc5996bc7388b803aad4233fd5582e0bad), where the output artifact can be found [here](https://github.com/PalisadoesFoundation/talawa/releases). We would like to complete the `iOS-Build` section of this pipeline to generate `.ipa` files for iOS devices. This is a non-trivial task compared to its .apk counterpart as Apple's infrastructure is far more tightly bound and requires a number of workarounds to do this.  Try to automate this process as much as possible following the instructions given [here](https://www.cobeisfresh.com/blog/how-to-implement-a-ci-cd-workflow-for-ios-using-github-actions). Ensure to document any roadblocks/obstacles at you encounter. 

**Describe the solution you'd like**
We would like to complete the `iOS-Build` section of this pipeline to generate `.ipa` files for iOS devices. This is a non-trivial task compared to its .apk counterpart as Apple's infrastructure is far more tightly bound and requires a number of workarounds to do this.  Try to automate this process as much as possible following the instructions given [here](https://www.cobeisfresh.com/blog/how-to-implement-a-ci-cd-workflow-for-ios-using-github-actions). Ensure to document any roadblocks/obstacles at you encounter.  At the end link this new output to the README.md of the repository.

### PR DESCRIPTION
**Pull request by Dante291**
**What kind of change does this PR introduce?**

Implements automated processes for iOS build and release in order to streamline development and deployment.

**Issue Number:**

Fixes #2136 

**Did you add tests for your changes?**

No

**Snapshots/Videos:**

image

**If relevant, did you update the documentation?**

No

**Does this PR introduce a breaking change?**

Enhancing CI/CD workflow for automation

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by Dante291** [comment]
@DMills27 @palisadoes 

Sir the [blog](https://www.cobeisfresh.com/blog/how-to-implement-a-ci-cd-workflow-for-ios-using-github-actions) provided by @DMills27 for the automation of build and release process actually need a paid apple developer account. 

 The current configuration for the iOS build in this workflow is set up to generate an .ipa file without code signing.
 This means that the generated .ipa file can only be run on a virtual iOS device (simulator). This setup is intended
 to automate the process of building the iOS app in a basic form.

 To build an iOS app with code signing, which is necessary for distribution and running the app on actual iOS devices, a paid service of the Apple Developer Program is required. This workflow does not include steps for code signing due to the 
 associated costs and the need for individual Apple Developer accounts.

**Comment by Dante291** [comment]
One solution for the above problem could be:

To facilitate testing on actual iOS devices for our open-source Flutter project, developers can follow a set of comprehensive steps outlined in the INSTALLATION.md file. These instructions include cloning the repository, installing dependencies (although already mentioned in INSTALLATION.md), configuring Xcode for code signing, registering their iOS device, and finally, building and running the app.

 This guide ensures that developers, with their own free Apple Developer account, can seamlessly set up their environment and test the app on their iOS device. It serves as a valuable resource in our project's documentation, providing a clear pathway for contributors interested in testing on real iOS devices.
 
 
 @DMills27 @palisadoes @noman2002 Your views on this solution?

**Comment by Dante291** [comment]
@palisadoes Could you assign reviewers for this, thanks.

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -157,3 +157,27 @@ jobs:
           architecture: x64
       - name: Building for ios
         run: flutter build ios --release --no-codesign
+        # '--no-codesign' is used for building without code signing.
```
Comment (literalEval) [inline .github/workflows/pull-request.yml:160]: Don't need build on Pull Request. Please remove from here.

**Review by literalEval — CHANGES_REQUESTED** [review]

**Comment by Dante291** [comment]
@literalEval Done sir.

File: .github/workflows/push.yaml [hunk 2]
```diff
@@ -235,3 +235,27 @@ jobs:
           architecture: x64
       - name: Building for ios
         run: flutter build ios --release --no-codesign
+        # '--no-codesign' is used for building without code signing.
```
Comment (literalEval) [inline .github/workflows/push.yaml:238]: Rename:

1. Workflow to 'ios-build-and-release' (with appropriate casing of course)
2. Built ipa to 'app-release.ipa' to ensure consistency with the apk.

**Review by literalEval — CHANGES_REQUESTED** [review]

File: .github/workflows/push.yaml (hunk 2 above)
Comment (Dante291) [reply .github/workflows/push.yaml:238]: @literalEval made changes sir.

Comment (literalEval) [reply .github/workflows/push.yaml:238]: 1. I meant the whole iOS Build workflow; not the specific one.

**Review by literalEval — APPROVED** [review]
LGTM.

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. How can we get the file added to this area or something similar?
2. This will allow us to point all developers to a single location to download the latest iOS and Android code

- https://github.com/PalisadoesFoundation/talawa/releases

I'd prefer a universal solution versus having people visit different locations for different mobile operating systems. Simplicity is always best.

**Comment by Dante291** [comment]
> One solution for the above problem could be:
> 
> To facilitate testing on actual iOS devices for our open-source Flutter project, developers can follow a set of comprehensive steps outlined in the INSTALLATION.md file. These instructions would include cloning the repository, installing dependencies (although already mentioned in INSTALLATION.md), configuring Xcode for code signing, registering their iOS device, and finally, building and running the app.
> 
> This guide will ensure that developers, with their own free Apple Developer account, can seamlessly set up their environment and test the app on their iOS device. It will serve as a valuable resource in our project's documentation, providing a clear pathway for contributors interested in testing on real iOS devices.
> 
> @DMills27 @palisadoes @noman2002 Your views on this solution?

@palisadoes we could go for this.

**Comment by Dante291** [comment]
> 1. How can we get the file added to this area or something similar?
> 2. This will allow us to point all developers to a single location to download the latest iOS and Android code
> 
> * https://github.com/PalisadoesFoundation/talawa/releases
> 
> I'd prefer a universal solution versus having people visit different locations for different mobile operating systems. Simplicity is always best.

The build iOS app will appear in the release section of the repo only, like this:

image

**Comment by literalEval** [comment]
@palisadoes 

1. Releases via this workflow run will go to the said [URL](https://github.com/PalisadoesFoundation/talawa/releases), so it's not a problem.
2. For the issue of having different files for different platforms, we can't really do anything. Android and iOS have different installers because these OSes are fundamentally different from each other.

**Comment by Dante291** [comment]
@palisadoes please review

**Comment by palisadoes** [comment]
@DMills27 Please take a look. You've worked on this too.

**Comment by DMills27** [comment]
Your solution is fine as-is; you need not concern yourself with the signing the .ipa files. I suggest you look into how [FlutterFlow](https://flutterflow.io/) automates this process for iPhone devices.

**Comment by Dante291** [comment]
@DMills27 Ok sir will learn about third party services for the automation of deployment of iOS app

**Comment by Dante291** [comment]
@palisadoes This can be merged now as said by @DMills27 that code signing is not required now.

**Comment by Dante291** [comment]
> Your solution is fine as-is; you need not concern yourself with the signing the .ipa files. I suggest you look into how [FlutterFlow](https://flutterflow.io/) automates this process for iPhone devices.
> 
> Also, leaving these links here for posterity:
> 
> https://discourse.nixos.org/t/using-nix-for-development-building-and-testing-android-project/191/8 https://sandervanderburg.blogspot.com/2012/12/deploying-ios-applications-with-nix.html https://medium.com/pinterest-engineering/continuous-integration-for-ios-with-nix-and-buildkite-ef5b36c5292d https://github.com/uniphil/nix-for-devs#react-native

@palisadoes So current solution for automation of deployment of iOS app without code signing is fine as stated by @DMills27 too so I guess this can be merged now.

**Review by palisadoes — APPROVED** [review]

**Pull request merged**